staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -300,3 +307,61 @@ func (le *LeaderElector) maybeReportTransition() {
 		go le.config.Callbacks.OnNewLeader(le.reportedLeader)
 	}
 }
+
+type LeaderChecker interface {
+	healthz.HealthzChecker
+	SetLeaderElection(le *LeaderElector)
+}
+
+type timeNow func() time.Time
+
+type leaderChecker struct {
+	c           *apiserverconfig.LeaderElectionConfiguration
+	pointerLock sync.Mutex
+	le          *LeaderElector
+	timeNow
+}
+
+func (l *leaderChecker) Name() string {
+	return "leaderElection/" + l.c.ResourceLock
+}
+
+func (l *leaderChecker) Check(req *http.Request) error {
+	if !l.c.LeaderElect {
+		return nil
+	}
+	l.pointerLock.Lock()
+	defer l.pointerLock.Unlock()
+	if l.le == nil {
+		return nil
+	}
+	if !l.le.IsLeader() {
+		// Currently not concerned with the case that we are hot standby
+		return nil
+	}
+	now := l.timeNow()
+	// If we are more than 20 seconds after the lease durection that is past the timeout on the lease renew.
+	// Time to start reporting ourselves as unhealthy. We should have died but conditions like deadlock can
+	// prevent this. (See #70819)
+	expiry := l.le.observedTime.Add(l.le.config.LeaseDuration).Add(time.Second * 20)
+
+	if now.After(expiry) {
+		return fmt.Errorf("Failed to renew leader election for %s", l.c.ResourceLock)
+	}
+
+	return nil
+}
+
+func (l *leaderChecker) SetLeaderElection(le *LeaderElector) {
+	l.pointerLock.Lock()
+	defer l.pointerLock.Unlock()
+	l.le = le
+}
+
+func GetLeaderHealthCheck(c *apiserverconfig.LeaderElectionConfiguration) LeaderChecker {
```

Comment:
I don't think we should be importing apiserverconfig here, this is a client library. Can you abstract something with an interface?

Reply:
Done.